doc/LanguageClient.txt

```diff
@@ -386,6 +386,124 @@ Maximum severity to show diagnostic messages.
 Default: "Hint"
 Valid options: "Error" | "Warning" | "Information" | "Hint"
 
+2.29 g:LanguageClient_semanticHighlightMaps   *g:LanguageClient_semanticHighlightMaps*
+
+String to list/map map. Defines the mapping of semantic highlighting "scopes" to
+highlight groups. This depends on the LSP server supporting the proposed
+semantic highlighting protocol, see:
+
+https://github.com/microsoft/language-server-protocol/issues/18
+https://github.com/microsoft/vscode-languageserver-node/issues/368
+
+
+Like |g:LanguageClient_serverCommands| this is a map where the keys are
+filetypes. However the value at each key can be either:
+
+1. A map of highlight group name (see |highlight-groups|) => list of strings:
+>
+    let g:LanguageClient_semanticHighlightMaps = {
+        \ 'java': {
+        \   "Function": ['entity.name.function.java', '**'],
+        \   "Type": ['entity.name.type.class.java', '**'],
+        \ }
+        \ }
+
+2. A list of maps from #1, this is to allow for multiple patterns for the same
+   highlight group.
+>
+    let g:LanguageClient_semanticHighlightMaps = {
+        \ 'java': [
+        \   {"Function": ['entity.name.function.java', '**']},
```

Comment:
Do you think if it's more intuitive if the map is mapping semantic groups to highlight groups?

Another related question, what is the motivation of supporting two styles of this setting? Should one suffice?

Reply:
The main issue/reason for the reversed map is that vimscript doesn't support arrays as a key into a dictionary.

Originally I just had a single map but I realized afterward that theres often a need to map multiple scopes to the same highlight group. I guess I could simplify the documentation and only mention the second style but still allow the first if it ever gets used.

Reply:
Another idea is to represent semantic scopes as a string with some separator, e.g. `storage.modifier.static.java/entity.name.function.java/**`. That would allow using them as keys. Although idk if there are any banned characters from the semantic scope names that would be usable as separators.

Reply:
Let's reduce the number of supported styles here. More styles supported only leads to complicated implementation to maintainer and more confusion to end user.

@YaLTeR cleaver idea! Agreed the tricky part is the choice of separator. As different language have different rules, I don't know there are possible separators working for all languages.

Reply:
Continuing the same approach presented by @YaLTeR. To solve the issue of ambiguity of separators, a buffer local variable with default can be introduced, say `b:LanguageClient_semanticScopeSeparator` with default of `/`.

e.g., user might have config like
```vim
let g:LanguageClient_semanticHighlightMaps['java'] = {
    \ 'storage.modifier.static.java/entity.name.function.java/.*': 'Function',
}
```

In a different filetype where `/` might be part of the scope themselves,
```vim
autocmd *.elisp setlocal LanguageClient_semanticScopeSeparator='|'
let g:LanguageClient_semanticHighlightMaps['elisp'] = {
    \ 'storage.modifier.static|name.function|.*': 'Function',
}
```

We would construct scope by concatenate the scopes with this separator. The actual highlight group are then being looked up through this map.

IMO, this way is more intuitive and solves the original issue that `**` can only appear in beginning or end.